File: src/aoc2024/day2.rs
use std::io::BufRead;

fn part_a(lines: &Vec<String>) {
    let values: Vec<Vec<i32>> = lines
        .iter()
        .map(|x| {
            x.split_whitespace()
                .map(|y| y.parse::<i32>().unwrap())
                .collect()
        })
        .collect();
    let differences: Vec<Vec<i32>> = values
        .into_iter()
        .map(|x| {
            let len = x.len();
            x.as_slice()[..len - 1]
                .iter()
                .zip(x.as_slice()[1..len].iter())
                .map(|(a, b)| b - a)
                .collect()
        })
        .collect();

    let result: i32 = differences
        .into_iter()
        .map(|x| {
            ((x.iter().all(|y| *y < 0) || x.iter().all(|y| *y > 0))
                && x.iter().all(|y| i32::abs(*y) <= 3)) as i32
        })
        .sum();

    println!("{result}");
}

fn part_b(lines: &Vec<String>) {
    let result: i32 = lines
        .iter()
        .map(|x| {
            x.split_whitespace()
                .map(|y| y.parse::<i32>().unwrap())
                .collect()
        })
        .map(|values: Vec<i32>| match evaluate(&values, None) {
            true => 1,
            false => {
                for i in 0..values.len() {
                    if evaluate(&values, Some(i)) {
                        return 1;
                    }
                }
                return 0;
            }
        })
        .sum();
    println!("{result}");
}

fn evaluate(values: &Vec<i32>, i: Option<usize>) -> bool {
    match i {
        None => {
            let len = values.len();
            let differences: Vec<i32> = values.as_slice()[..len - 1]
                .iter()
                .zip(values.as_slice()[1..len].iter())
                .map(|(a, b)| *a - *b)
                .collect();
            (differences.iter().all(|y| *y < 0) || differences.iter().all(|y| *y > 0))
                && differences.iter().all(|y| i32::abs(*y) <= 3)
        }
        Some(j) => {
            let values: Vec<i32> = values
                .into_iter()
                .enumerate()
                .filter(|(i, _)| *i != j)
                .map(|(_, x)| *x)
                .collect();
            let len = values.len();
            let differences: Vec<i32> = values.as_slice()[..len - 1]
                .iter()
                .zip(values.as_slice()[1..len].iter())
                .map(|(a, b)| *a - *b)
                .collect();
            (differences.iter().all(|y| *y < 0) || differences.iter().all(|y| *y > 0))
                && differences.iter().all(|y| i32::abs(*y) <= 3)
        }
    }
}

pub fn run() {
    let lines: Vec<String> = std::io::stdin()
        .lock()
        .lines()
        .map(|x| x.unwrap())
        .collect();

    part_a(&lines);
    part_b(&lines);
}

File: python/src/aoc2024/day2.py
import sys


def part_a(lines: list[str]):
    result = 0
    for line in lines:
        values = [int(x) for x in line.split()]
        differences = [a - b for a, b in zip(values[1:], values[:-1])]

        result += (
            all([x < 0 for x in differences]) or all([x > 0 for x in differences])
        ) and all([abs(x) <= 3 for x in differences])

    print(result)


def part_b(lines: list[str]):
    result = 0
    for line in lines:
        values = [int(x) for x in line.split()]

        if evaluate(values):
            result += 1
        else:
            for i in range(len(values)):
                if evaluate(values, i):
                    result += 1
                    break

    print(result)


def evaluate(values: list[int], i=None):
    if i is not None:
        values = [x for j, x in enumerate(values) if j != i]

    differences = [a - b for a, b in zip(values[1:], values[:-1])]
    return (
        all([x < 0 for x in differences]) or all([x > 0 for x in differences])
    ) and all([abs(x) <= 3 for x in differences])


def run():
    lines = [line.strip() for line in sys.stdin]
    part_a(lines)
    part_b(lines)


if __name__ == "__main__":
    run()
